File: src/main/java/io/github/shaksternano/borgar/command/ChangeExtensionCommand.java
package io.github.shaksternano.borgar.command;

import com.google.common.collect.ListMultimap;
import io.github.shaksternano.borgar.exception.UnsupportedFileFormatException;
import io.github.shaksternano.borgar.io.FileUtil;
import io.github.shaksternano.borgar.io.NamedFile;
import net.dv8tion.jda.api.events.message.MessageReceivedEvent;

import java.io.File;
import java.io.IOException;
import java.util.List;

public class ChangeExtensionCommand extends FileCommand {

    private final String newExtension;

    public ChangeExtensionCommand(String newExtension) {
        super(newExtension + 2, "Changes the extension of a file to `." + newExtension.toLowerCase() + "`.");
        this.newExtension = newExtension;
    }

    @Override
    protected NamedFile modifyFile(File file, String fileName, String fileFormat, List<String> arguments, ListMultimap<String, String> extraArguments, MessageReceivedEvent event, long maxFileSize) throws IOException {
        if (FileUtil.getFileExtension(fileName).equals(newExtension)) {
            throw new UnsupportedFileFormatException("The file already has the extension `." + newExtension.toLowerCase() + "`!");
        }
        return FileUtil.changeFileExtension(file, fileName, newExtension);
    }
}

File: core/src/main/kotlin/io/github/shaksternano/borgar/core/command/ChangeExtensionCommand.kt
package io.github.shaksternano.borgar.core.command

import com.google.common.collect.ListMultimap
import com.google.common.io.Files
import io.github.shaksternano.borgar.core.command.util.CommandResponse
import io.github.shaksternano.borgar.core.util.MessageUtil
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.engine.cio.*
import io.ktor.client.request.*
import io.ktor.http.*
import kotlinx.coroutines.future.await
import net.dv8tion.jda.api.entities.Message
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import java.io.InputStream
import kotlin.jvm.optionals.getOrElse

class ChangeExtensionCommand(
    private val newExtension: String
) : KotlinCommand<Unit>(
    "${newExtension}2",
    "Changes the extension of a file to `.$newExtension`.",
) {

    override suspend fun executeSuspend(
        arguments: List<String>,
        extraArguments: ListMultimap<String, String>,
        event: MessageReceivedEvent
    ): CommandResponse<Unit> {
        val url = MessageUtil.getUrlTenor(event.message)
            .await()
            .getOrElse {
                return CommandResponse("No media found!")
            }
        val urlNoQueryParams = url.split('?').first()
        val oldExtension = Files.getFileExtension(urlNoQueryParams)
        if (oldExtension == newExtension) {
            return CommandResponse("File already has the extension `.$newExtension`!")
        }
        HttpClient(CIO).use { client ->
            val response = client.get(url)
            val contentLength = response.contentLength() ?: 0
            if (contentLength > Message.MAX_FILE_SIZE) {
                return CommandResponse("File is too large!")
            }
            val fileNameWithoutExtension = Files.getNameWithoutExtension(urlNoQueryParams)
            val fileName =
                if (newExtension.isBlank()) fileNameWithoutExtension
                else "$fileNameWithoutExtension.$newExtension"
            val inputStream = response.body<InputStream>()
            return CommandResponse(
                inputStream,
                fileName
            )
        }
    }
}
